### ISSUE TITLE
Consider support for rand 0.7

### ISSUE THREAD

**Issue by Darksonn** [rust-num/num-bigint#75]
Currently the `rand` feature of `num-bigint` uses `rand` version 0.5, although version 0.6 has been released.  According to the rand documentation the update shouldn't affect the required version of rustc for the `rand` feature.

I can't quite tell what effects it would have for backwards compatibility in general, but maybe one could use a range for the version.

Related issue: [#47][1]

  [1]: https://github.com/rust-num/num-bigint/issues/47

**Comment by cuviper** [comment]
Just trying the update, I found that the method signatures of [`UniformSampler`](https://docs.rs/rand/0.6.4/rand/distributions/uniform/trait.UniformSampler.html) have changed from [0.5](https://docs.rs/rand/0.5.5/rand/distributions/uniform/trait.UniformSampler.html), which makes it harder to support a version range here. _Maybe_ that could be detected in the build script for a `cfg` flag.

Generally though, I think the experience with a version range was poor in the past, as Cargo really tries to just grab the newest possible.  And if one part of your dependency tree ends up wanting num-bigint+rand-0.6, while another wants num-bigint+rand-0.5, then you're stuck.

**Comment by cuviper** [comment]
Rand 0.7 just came out today.

I think what I may try is bumping to num-bigint 0.3 with support for rand 0.7, and then have a num-bigint 0.2.x update which re-exports everything from 0.3 except the rand support, keeping its own support for rand 0.5.

(Yes, this means skipping 0.6 -- I don't want to version lock with their every update.)

**Comment by cuviper** [comment]
Oh, right, we can't do a "re-export and add rand 0.5" in 0.2.x due to coherence:

```
error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
   --> src/bigrand.rs:148:1
    |
148 | / impl SampleUniform for BigUint {
149 | |     type Sampler = UniformBigUint;
150 | | }
    | |_^ impl doesn't use types inside crate
    |
    = note: the impl does not reference any types defined in this crate
    = note: define and implement a trait or new type instead
```

I think everything else would work, but `SampleUniform` is a public `impl`, so that's a blocker.

So nevermind, it will have to be a fully independent bump.

**Comment by programmerjake** [comment]
Maybe depend on both versions of rand and allow selecting which one(s) to use using features?
I don't know if rust can handle using two separate versions of the same crate at the same time (maybe give them different crate-internal aliases).

**Comment by programmerjake** [comment]
also, maybe rename the issue to "Consider support for rand 0.7"

**Comment by danclive** [comment]
I also encountered this problem today, because I am currently using 'rand v0.7', using this' crate 'will encounter conflicts, so I can only fork one copy, and then manually change the' rand 'version to' v0.7 '.

**Comment by YorickPeterse** [comment]
Has any progress on this been made? It would be nice being able to use newer versions of rand combined with the random number generation support provided by num-bigint.

**Comment by cuviper** [comment]
#110 just merged with rand 0.7 as part of the upcoming num-bigint 0.3 -- #128.

**Comment by YorickPeterse** [comment]
@cuviper @kpcyrd Thanks for adding rand 0.7 support!
